File: solution/python3/347.top-k-frequent-elements.py
#
# @lc app=leetcode id=347 lang=python3
#
# [347] Top K Frequent Elements
#

# @lc code=start
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        heap = []
        tmp = {}
        count = 0
        nums.sort()
        for i in range(len(nums)):
            count += 1
            if i == len(nums) - 1 or nums[i] != nums[i+1]:
                heapq.heappush(heap, count)
                if count in tmp:
                    tmp[count].append(nums[i])
                else:
                    tmp[count] = [nums[i]]
                count = 0

        print(tmp)
        ans = []
        for i in heapq.nlargest(k, heap):
            ans.extend(tmp[i])
        return set(ans)

# @lc code=end

File: solution/java8/347.top-k-frequent-elements.java
/*
 * @lc app=leetcode id=347 lang=java
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
import java.util.*;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> counter = new HashMap<>();
        for (int i: nums) {
            if (counter.containsKey(i)) {
                int val = counter.get(i);
                counter.put(i, val + 1);
            } else {
                counter.put(i, 1);
            }
        }

        PriorityQueue<Map.Entry<Integer, Integer>> queue
            = new PriorityQueue<>(Map.Entry.comparingByValue(Comparator.reverseOrder()));
        for (Map.Entry<Integer, Integer> entry: counter.entrySet()) {
            queue.add(entry);
        }

        int[] ans = new int[k];
        for (int i = 0; i < k; i++) {
            Map.Entry<Integer, Integer> entry = queue.poll();
            ans[i] = entry.getKey();
        }
        return ans;
    }
}
// @lc code=end

